### PR TITLE
fix(dropdown): container body keyboard navigation

### PR DESCRIPTION
**Pull request by benouat**
This PR is a follow up PR to #3625 which introduced the ability to close the dropdown when leaving it by the last elements.
Unfortunately, it did not fix the keyboard navigation when using `container="body"` is used.

Leaving the dropdown menu from the bottom when container is set to "body" leaves the user in a broken state, where the browser window url bar gets the focus.

The PR introduces a mechanism to _catch_ the last <kbd>Tab</kbd> from the dropdown menu, and temporary set the focus back in a synchronous fashion to the dropdown anchor, hence letting the browser placing the focus on the next focusable element.

It also makes sure now that whenever the user wants to close the dropdown from the keyboard with <kbd>Esc</kbd>, the focus gets back to the dropdown toggle element.

This used to be the broken behaviour that now works
Kapture 2020-06-29 at 11 44 04

### CODE REVIEW

File: src/dropdown/dropdown.ts [hunk 1]
```diff
@@ -324,16 +334,30 @@ export class NgbDropdown implements AfterContentInit, OnDestroy {
...
-          itemElements[0].focus();
-          event.preventDefault();
-        } else if (this.container === 'body' && event.shiftKey && position === 0) {
-          this._anchor.anchorEl.focus();
-          event.preventDefault();
+        if (this._anchor.anchorEl === event.target) {
+          if (this.container === 'body' && !event.shiftKey) {
+            this._renderer.setAttribute(this._menuElement.nativeElement, 'tabindex', '0');
+            this._menuElement.nativeElement.focus();
+            this._renderer.removeAttribute(this._menuElement.nativeElement, 'tabindex');
```
[the remark this answers is not in the document]
Comment (benouat) [reply src/dropdown/dropdown.ts:341]: We don't want the menu to be permanently focusable. It would make reachable with keyboard, which we clearly don't won't.

The trick here is to be able to _find_ the next focusable element in the menu, mostly without to specify it. Previous PR #3625 was wrongly assuming that focusable elements were only `ngbDropdownItem`. Simply have a look at the [form demo](https://ng-bootstrap.github.io/#/components/dropdown/examples#navbar)

To do that, I listen on the `keydown.Tab` from the menu anchor/toggle element. And when we are about to enter the menu, I temporarily and synchronously focus the menu, so that the browser will natively continue to forward the `Tab` **inside** the menu, hence focusing whatever focusable coming first.

Do you want me to comment it out in the code ?

File: src/dropdown/dropdown.ts [hunk 2]
```diff
@@ -324,16 +334,30 @@ export class NgbDropdown implements AfterContentInit, OnDestroy {
...
+            this.close();
+          }
+          return;
+        } else if (this.container === 'body') {
+          const focusableElements = [...this._menuElement.nativeElement.querySelectorAll(FOCUSABLE_ELEMENTS_SELECTOR)];
+          if (event.shiftKey && event.target === focusableElements[0]) {
+            this._anchor.anchorEl.focus();
+            event.preventDefault();
+          } else if (!event.shiftKey && event.target === focusableElements[focusableElements.length - 1]) {
+            this._anchor.getNativeElement().focus();
```
[the remark this answers is not in the document]
Comment (benouat) [reply src/dropdown/dropdown.ts:352]: Bad copy paste here from the old PR.
Which one do we keep ? I think I don't understand the need of `getNativeElement()` when `anchorEl` is publicly available

File: src/dropdown/dropdown.ts [hunk 3]
```diff
@@ -249,14 +250,23 @@ export class NgbDropdown implements AfterContentInit, OnDestroy {
       this._applyContainer(this.container);
       this.openChange.emit(true);
       this._setCloseHandlers();
+      if (this._anchor) {
```
[the remark this answers is not in the document]
Comment (benouat) [reply src/dropdown/dropdown.ts:253]: I think I prefer a good old `if`. Not very keen of TypeScript `?.` syntax.

[the remark this answers is not in the document]
Comment (pkozlowski-opensource) [reply src/dropdown/dropdown.ts:253]: FYI, this is stage 4 ECMAScript proposal, so not TS specific really: https://github.com/tc39/proposal-optional-chaining

[the remark this answers is not in the document]
Comment (benouat) [reply src/dropdown/dropdown.ts:253]: > FYI, this is stage 4 ECMAScript proposal, so not TS specific really: https://github.com/tc39/proposal-optional-chaining

Already stage 4! Wow.

Then, let me rephrase it: _not very keen of `?.` syntax when used with function calls._

I might be weird on the topic, but to me, optional chaining makes sense, and particularly shines when you access/read properties. It does not click to me to use it when the property is an actual function... Don't ask me why, I don't know.

```typescript
const baz = foo?.bar.baz; // sounds ok to me
foo?.bar.baz(); // sounds weird to me
```

But as you both jumped on the topic, I can change it... no big deal.

[the remark this answers is not in the document]
Comment (pkozlowski-opensource) [reply src/dropdown/dropdown.ts:253]: > foo?.bar.baz(); // sounds weird to me

@benouat actually, I DO agree with you! Personally I would also _avoid_ using ?. in this particular case - but I guess a matter of preference. I just commented to note that this is not TS-specific.

[the remark this answers is not in the document]
Comment (maxokorokov) [reply src/dropdown/dropdown.ts:253]: @benouat I agree that it's pretty subjective, so keep it as is

File: src/dropdown/dropdown.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (maxokorokov) [reply src/dropdown/dropdown.ts:352]: > I think I don't understand the need of `getNativeElement()` when `anchorEl` is publicly available

That was my point too, need to track history, but I don't think it's useful

File: src/dropdown/dropdown.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (maxokorokov) [reply src/dropdown/dropdown.ts:341]: I see now, please do add a short comment

**Comment by benouat** [comment]
I created a new commit regarding the refactoring of how to access `nativeElement` across both `anchor` and `menu` so it could be easier to review.

**Comment by benouat** [comment]
> I guess the second commit should be `refactor`, not `fix`, but I'm not sure if you want to squash them in the end or not.

I'll squash it!
